### DISCUSSION TITLE
Question about data access for Parse Biosciences Sketch Integration tutorial

### DISCUSSION THREAD

**Discussion by mbm2224** [satijalab/seurat#7511] [Q&A]
Hello, I have a question about how the data were accessed in [Sketch integration using a 1 million cell dataset from Parse Biosciences vignette](https://satijalab.org/seurat/articles/parsebio_sketch_integration). The tutorial states that the dataset from[ Parse Biosciences](https://resources.parsebiosciences.com/dataset-wt-mega-one-million-pbmc-type-1-diabetes) is downloaded as an h5.ad file.

I followed the link to [Using BPCells with Seurat Objects vignette](https://satijalab.org/seurat/articles/seurat5_bpcells_interaction_vignette.html#parse-biosciences) to see how the data were downloaded and processed. The download link for the data takes me to [a Parse Biosciences webpage](https://support.parsebiosciences.com/hc/en-us/articles/7704577188500-How-to-analyze-a-1-million-cell-data-set-using-Scanpy-and-Harmony). There are instructions to download the genes, matrix, and metadata file, but no option for an h5.ad file. 

How was this file: 'ParseBio_PBMC.h5ad' obtained for this line of code 
```
parse.data <- open_matrix_anndata_hdf5(
  "/brahms/hartmana/vignette_data/h5ad_files/ParseBio_PBMC.h5ad"
```
 from [Using BPCells with Seurat Objects vignette](https://satijalab.org/seurat/articles/seurat5_bpcells_interaction_vignette.html#parse-biosciences)? Did you follow the steps for preprocessing from Parse Biosciences? If so, how far in the tutorial did you go until you wrote out the h5.ad file?

Thanks in advance for clarifying!

**Comment by mdidish** [comment]
Hello, have youf foud the answer?

I tried this:
1st with Python:
```
import scanpy as sc
adata = sc.read_mtx('DGE_1M_PBMC.mtx')
adata.write('ParseBio_PBMC.h5ad')
```
2nd with R
```
parse.data <- open_matrix_10x_hdf5(
  path = "../RawData/ParseBio_PBMC.h5ad"
)
```

But I obtained this error:
`Loading multi-genome matrices from old-style 10x hdf5 files is unsupported`

**Comment by ashwini-girish** [comment]
In case you are still looking for the dataset, you can download it using this [link](https://cdn.parsebiosciences.com/1M_PBMC_T1D_Parse.zip). The same is now linked in the article as well.

**Reply by mbm2224** [reply]
Thank you!

**Comment by Pancreas-Pratik** [comment]
For reference, I was having some trouble for re-creating the initial dataset used in the Parse Biosciences Sketch Integration tutorial, but this is how I was able to do it:

From the Seurat vignette section OP mentioned (https://satijalab.org/seurat/articles/seurat5_bpcells_interaction_vignette.html#parse-biosciences), I clicked the **here** where in that section it says: "The data is available for download [here](https://support.parsebiosciences.com/hc/en-us/articles/7704577188500-How-to-analyze-a-1-million-cell-data-set-using-Scanpy-and-Harmony).

I then did this part of the python code:

```
import os
import sys
import scipy
import numpy as np
import pandas as pd
import scanpy as sc
import seaborn as sns
import scipy.io as sio
import scanpy.external as sce
import matplotlib.pyplot as plt

sc.settings.verbosity = 1
sc.settings.figdir = '/my-figure-directory/'
obj_save_path = '/my-object-save-path/'
mat_path = '/my-expression-matrix-path/'

# Adjust Scanpy figure defaults
sc.settings.set_figure_params(dpi=100, fontsize=10, dpi_save=400,
    facecolor = 'white', figsize=(6,6), format='png')
```

and also this part, but instead of writing the `adata_obj1.h5ad` immediately after the `sc.read_mtx()` line. I moved it after the adding cell meta data. So basically the resulting .h5ad now also has cell and gene names.

```
# The DGE_filtered folder contains the expression matrix, genes, and files
adata = sc.read_mtx(mat_path + 'DGE_1M_PBMC.mtx')


# adata = sc.read(obj_save_path + 'adata_obj1.h5ad')

# reading in gene and cell data
gene_data = pd.read_csv(mat_path + 'all_genes_1M_PBMC.csv')
cell_meta = pd.read_csv(mat_path + 'cell_metadata_1M_PBMC.csv')

# find genes with nan values and filter
gene_data = gene_data[gene_data.gene_name.notnull()]
notNa = gene_data.index
notNa = notNa.to_list()

# remove genes with nan values and assign gene names
adata = adata[:,notNa]
adata.var = gene_data
adata.var.set_index('gene_name', inplace=True)
adata.var.index.name = None
adata.var_names_make_unique()

# add cell meta data to anndata object
adata.obs = cell_meta
adata.obs.set_index('bc_wells', inplace=True)
adata.obs.index.name = None
adata.obs_names_make_unique()

adata.write(obj_save_path + 'adata_obj1.h5ad')
```

and then loaded this in R using most of the code in the section (https://satijalab.org/seurat/articles/seurat5_bpcells_interaction_vignette.html#parse-biosciences), but I had to make some changes, basically I had to load in the meta data .csv, and ensure it was in a format to be used within the `CreateSeuratObject()` command. I _did_ do some checks myself to make sure this was being done correctly (which was just checking the order of cell names in the meta data and making sure it matched the order of the cells in the Seurat object)... make sure paths are correct, though, I replaced my path with just `~`... note: also `ParseBio_PBMC.h5ad` is named `adata_obj1.h5ad` here, but it's the same... :

```
parse.data <- open_matrix_anndata_hdf5("~/adata_obj1.h5ad")
write_matrix_dir(mat = parse.data, dir = "~/parse_1m_pbmc_example")
parse.mat <- open_matrix_dir(dir = "~/parse_1m_pbmc_example")
parse.meta_data_csv <- read.csv("/cell_metadata_1M_PBMC.csv")
object <- CreateSeuratObject(counts = parse.mat, meta.data = parse.meta_data_csv)
```

and then from here I continued with the Parse Biosciences Sketch Integration tutorial, here:
```
object <- NormalizeData(object)
object[["RNA"]] <- split(object[["RNA"]], f = object$sample)
so on and so forth... 
```

**Reply by emilykibbler** [reply]
I know this is an old thread but maybe someone will see this… is there another way to get the h5 file? I don’t have enough memory on my laptop or google colab to run that python script. The mtx file is too big. It looks like a good vignette and I would like to follow it

**Comment by mdidish** [comment]
Hello
Here is what I did: 

```
library(Seurat)
library(Matrix)

# set this option when analyzing large datasets
options(future.globals.maxSize = 4000 * 1024^2)

# read metadata
metadata <- read.table(file = "../RawData/1M_PBMC_T1D_Parse/cell_metadata_1M_PBMC.csv", header = T, sep = "\t")
rownames(metadata) <- metadata$bc_wells

# read features names
gene_names <- read.table(file = "../RawData/1M_PBMC_T1D_Parse/all_genes_1M_PBMC.csv", header = T, sep = "\t")

# read matrix data
matrix.path <- paste0("../RawData/1M_PBMC_T1D_Parse/DGE_1M_PBMC.mtx") 
mat <- readMM(file = matrix.path)
rownames(mat) <- metadata$bc_wells
colnames(mat) <- gene_names$gene_id

# matrix with 3000 cells for each selected sample 
cells_in_sample <- metadata %>%
  filter(sample %in% c("D_497", "D_498", "D_500", "D_501", "D_502", "H_3128", "H_409", "H_6625", "H_7108", "H_120")) %>%
  group_by(sample) %>%
  sample_n(3000) %>%
  pull(bc_wells)

# subset matrix and metadata
mat.subset <- mat[cells_in_sample, ]
metadata.subset <- metadata[cells_in_sample, ]
save(mat.subset, metadata.subset, file = "subset.RData")

# restart R to clean memory
load("subset.RData")
# Transpose the data and convert to sparse matrix.
#mat.subset <- as(t(as.matrix(mat.subset)), "sparseMatrix")

# Create Seurat Object
pbmc.hashtag <- CreateSeuratObject(counts = as.matrix(t(mat.subset)), meta.data = metadata.subset, min.cells = 3)
```

Best

**Reply by emilykibbler** [reply]
The readMM function fives me "vector memory limit of 16.0 Gb reached." The file is 16.68 Gb... so close :'(

EDIT: I was able to run it in colab on a highvRAM session (with small edits, fixing to read.csv instead of read.tsv)

For anyone who finds this thread in the future, here is the output of the above code in rds format: 
[google drive file link](https://drive.google.com/file/d/1k1WfgxeHqD6hLRj37jTMjdbq9MVrPADs/view?usp=sharing)
